File: web/src/diagnostics.ts
import { logger, LogEntry, LogLevel } from './logger';

export interface CapabilitiesReport {
    pressure: boolean;
    tilt: boolean;
    hover: boolean;
    eraser: boolean;
    refreshModes: string[];
    hardwareAcceleration: boolean;
}

export interface DiagnosticsReport {
    version: string;
    platform: string;
    userAgent: string;
    devicePixelRatio: number;
    screenResolution: string;
    hasPointerEvents: boolean;
    maxTouchPoints: number;
    selectedBackend: string;
    availableBackends: string[];
    fallbackReason: string | null;
    capabilities: CapabilitiesReport;
    refreshMode: string;
    recentLogs: LogEntry[];
}

export function collectDiagnostics(activeBackend?: string, isNativeSupported?: boolean): DiagnosticsReport {
    const isBrowser = typeof window !== 'undefined';
    const ua = isBrowser ? window.navigator.userAgent : 'Node.js / Server Environment';
    const dpr = isBrowser ? window.devicePixelRatio || 1 : 1;
    const res = isBrowser ? `${window.screen?.width || 0}x${window.screen?.height || 0}` : '0x0';
    const hasPointer = isBrowser && typeof window.PointerEvent !== 'undefined';
    const maxTouch = isBrowser ? window.navigator.maxTouchPoints || 0 : 0;

    const backend = activeBackend || (isNativeSupported ? 'NativeOpenInkBridge' : 'PointerEventFallback');
    const fallbackReason = isNativeSupported
        ? null
        : 'OpenInkBridgeNative bridge object not detected on window; using HTML5 PointerEvents fallback';

    return {
        version: '0.1.2',
        platform: 'Web SDK',
        userAgent: ua,
        devicePixelRatio: dpr,
        screenResolution: res,
        hasPointerEvents: hasPointer,
        maxTouchPoints: maxTouch,
        selectedBackend: backend,
        availableBackends: ['NativeOpenInkBridge', 'PointerEventFallback', 'WasmCore'],
        fallbackReason: fallbackReason,
        capabilities: {
            pressure: true,
            tilt: true,
            hover: true,
            eraser: true,
            refreshModes: ['Fast', 'Partial', 'Full', 'Clear'],
            hardwareAcceleration: !!isNativeSupported
        },
        refreshMode: isNativeSupported ? 'Fast' : 'Software',
        recentLogs: logger.getRingBufferLogs()
    };
}

export function dumpConfiguration(activeBackend?: string, isNativeSupported?: boolean): string {
    const diag = collectDiagnostics(activeBackend, isNativeSupported);
    let out = '';
    out += '========== OpenInkBridge Diagnostics ==========\n';
    out += `Version: ${diag.version}\n`;
    out += `Platform: ${diag.platform}\n`;
    out += `User Agent: ${diag.userAgent}\n`;
    out += `Screen: ${diag.screenResolution} (DPR: ${diag.devicePixelRatio})\n`;
    out += `Pointer Events: ${diag.hasPointerEvents ? 'Supported' : 'Unsupported'} (Max Touch: ${diag.maxTouchPoints})\n`;
    out += `Selected Backend: ${diag.selectedBackend}\n`;
    out += `Available Backends: ${diag.availableBackends.join(', ')}\n`;
    if (diag.fallbackReason) {
        out += `Fallback Reason: ${diag.fallbackReason}\n`;
    }
    out += 'Capabilities:\n';
    out += `  - Pressure: ${diag.capabilities.pressure ? 'Supported' : 'Unsupported'}\n`;
    out += `  - Tilt: ${diag.capabilities.tilt ? 'Supported' : 'Unsupported'}\n`;
    out += `  - Hover: ${diag.capabilities.hover ? 'Supported' : 'Unsupported'}\n`;
    out += `  - Eraser: ${diag.capabilities.eraser ? 'Supported' : 'Unsupported'}\n`;
    out += `  - Refresh Modes: [${diag.capabilities.refreshModes.join(', ')}]\n`;
    out += `  - Hardware Acceleration: ${diag.capabilities.hardwareAcceleration ? 'Enabled' : 'Disabled'}\n`;
    out += `Refresh Mode: ${diag.refreshMode}\n`;
    out += '===============================================\n';
    return out;
}

export function createBugReport(activeBackend?: string, isNativeSupported?: boolean): string {
    let out = dumpConfiguration(activeBackend, isNativeSupported);
    out += '\n========== Recent Warnings & Errors ==========\n';
    const warnErrorLogs = logger.getRingBufferLogs().filter(e => e.level === LogLevel.WARN || e.level === LogLevel.ERROR);

    if (warnErrorLogs.length === 0) {
        out += 'No warnings or errors reported in recent log buffer.\n';
    } else {
        for (const entry of warnErrorLogs) {
            const levelStr = LogLevel[entry.level];
            out += `[${levelStr}][${entry.subsystem}][${entry.backend}] ${entry.event}: ${entry.message}\n`;
        }
    }
    out += '===============================================\n';
    return out;
}

File: core/src/diagnostics.rs
use crate::logging::{LogEntry, LogLevel, get_ring_buffer_entries};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Capabilities {
    pub pressure: bool,
    pub tilt: bool,
    pub hover: bool,
    pub eraser: bool,
    pub refresh_modes: Vec<String>,
    pub hardware_acceleration: bool,
}

impl Default for Capabilities {
    fn default() -> Self {
        Self {
            pressure: true,
            tilt: true,
            hover: true,
            eraser: true,
            refresh_modes: vec![
                "Fast".to_string(),
                "Partial".to_string(),
                "Full".to_string(),
                "Clear".to_string(),
            ],
            hardware_acceleration: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiagnosticsReport {
    pub version: String,
    pub platform: String,
    pub os: String,
    pub device_model: String,
    pub selected_backend: String,
    pub available_backends: Vec<String>,
    pub fallback_reason: Option<String>,
    pub capabilities: Capabilities,
    pub refresh_mode: String,
    pub build_configuration: String,
    pub feature_flags: Vec<String>,
    pub recent_logs: Vec<LogEntry>,
}

pub fn detect_platform() -> &'static str {
    if cfg!(target_arch = "wasm32") {
        "WASM / Web Browser"
    } else if cfg!(target_os = "android") {
        "Android"
    } else if cfg!(target_os = "linux") {
        "Linux"
    } else {
        "Unknown Platform"
    }
}

pub fn detect_feature_flags() -> Vec<String> {
    let mut flags = Vec::new();
    if cfg!(feature = "wasm") {
        flags.push("wasm".to_string());
    }
    if cfg!(feature = "android") {
        flags.push("android".to_string());
    }
    if cfg!(feature = "remarkable") {
        flags.push("remarkable".to_string());
    }
    flags
}

pub fn collect_diagnostics(
    selected_backend: String,
    available_backends: Vec<String>,
    fallback_reason: Option<String>,
    capabilities: Capabilities,
    refresh_mode: String,
) -> DiagnosticsReport {
    let build_config = if cfg!(debug_assertions) {
        "Debug"
    } else {
        "Release"
    };

    DiagnosticsReport {
        version: env!("CARGO_PKG_VERSION").to_string(),
        platform: detect_platform().to_string(),
        os: std::env::consts::OS.to_string(),
        device_model: std::env::var("OPENINKBRIDGE_DEVICE")
            .unwrap_or_else(|_| "Generic Device".to_string()),
        selected_backend,
        available_backends,
        fallback_reason,
        capabilities,
        refresh_mode,
        build_configuration: build_config.to_string(),
        feature_flags: detect_feature_flags(),
        recent_logs: get_ring_buffer_entries(),
    }
}

pub fn dump_configuration(report: &DiagnosticsReport) -> String {
    let mut out = String::new();
    out.push_str("========== OpenInkBridge Diagnostics ==========\n");
    out.push_str(&format!("Version: {}\n", report.version));
    out.push_str(&format!("Platform: {}\n", report.platform));
    out.push_str(&format!("OS: {}\n", report.os));
    out.push_str(&format!("Device Model: {}\n", report.device_model));
    out.push_str(&format!("Selected Backend: {}\n", report.selected_backend));
    out.push_str(&format!(
        "Available Backends: {}\n",
        report.available_backends.join(", ")
    ));

    if let Some(ref reason) = report.fallback_reason {
        out.push_str(&format!("Fallback Reason: {}\n", reason));
    }

    out.push_str("Capabilities:\n");
    out.push_str(&format!(
        "  - Pressure: {}\n",
        if report.capabilities.pressure {
            "Supported"
        } else {
            "Unsupported"
        }
    ));
    out.push_str(&format!(
        "  - Tilt: {}\n",
        if report.capabilities.tilt {
            "Supported"
        } else {
            "Unsupported"
        }
    ));
    out.push_str(&format!(
        "  - Hover: {}\n",
        if report.capabilities.hover {
            "Supported"
        } else {
            "Unsupported"
        }
    ));
    out.push_str(&format!(
        "  - Eraser: {}\n",
        if report.capabilities.eraser {
            "Supported"
        } else {
            "Unsupported"
        }
    ));
    out.push_str(&format!(
        "  - Refresh Modes: [{}]\n",
        report.capabilities.refresh_modes.join(", ")
    ));
    out.push_str(&format!(
        "  - Hardware Acceleration: {}\n",
        if report.capabilities.hardware_acceleration {
            "Enabled"
        } else {
            "Disabled"
        }
    ));

    out.push_str(&format!("Refresh Mode: {}\n", report.refresh_mode));
    out.push_str(&format!(
        "Build Configuration: {}\n",
        report.build_configuration
    ));
    out.push_str(&format!(
        "Feature Flags: [{}]\n",
        report.feature_flags.join(", ")
    ));
    out.push_str("===============================================\n");
    out
}

pub fn create_bug_report(report: &DiagnosticsReport) -> String {
    let mut out = dump_configuration(report);
    out.push_str("\n========== Recent Warnings / Errors ==========\n");
    let filtered_logs: Vec<&LogEntry> = report
        .recent_logs
        .iter()
        .filter(|e| e.level == LogLevel::Warn || e.level == LogLevel::Error)
        .collect();

    if filtered_logs.is_empty() {
        out.push_str("No warnings or errors reported in recent log buffer.\n");
    } else {
        for log in filtered_logs {
            out.push_str(&format!("{}\n", log.format_line()));
        }
    }
    out.push_str("===============================================\n");
    out
}
